eng/New-TestResources.ps1

```diff
@@ -0,0 +1,305 @@
+
+# Tag the resource group to be deleted after a certain number of hours if specified.
+$tags = @{
+    ServiceDirectory = $ServiceDirectory
+}
+
+if ($PSBoundParameters.ContainsKey('DeleteAfterHours')) {
+    $deleteAfter = [DateTime]::UtcNow.AddHours($DeleteAfterHours)
+    $tags.Add('DeleteAfter', $deleteAfter.ToString('o'))
+}
+
+if ($env:SYSTEM_TEAMPROJECTID) {
+    # Add tags for the current CI job.
+    $tags += @{
+        BuildId = "${env:BUILD_BUILDID}"
+        BuildJob = "${env:AGENT_JOBNAME}"
+        BuildNumber = "${env:BUILD_BUILDNUMBER}"
+        BuildReason = "${env:BUILD_REASON}"
+    }
+
+    # Set the resource group name variable.
+    Write-Host "##vso[task.setvariable variable=AZURE_RESOURCEGROUP_NAME;]$resourceGroupName"
+}
+
+Log "Creating resource group '${resourceGroupName}' in location '$Location'"
+$resourceGroup = New-AzResourceGroup -Name "${resourceGroupName}" -Location $Location -Tag $tags -Force:$Force
+if ($resourceGroup.ProvisioningState -eq 'Succeeded') {
+    # New-AzResourceGroup would've written an error and stopped the pipeline by default anyway.
+    Write-Verbose "Successfully created resource group '$($resourceGroup.ResourceGroupName)'"
+}
+
+# Populate the template parameters and merge any additional specified.
+$templateParameters = @{
+    baseName = $BaseName
+    testApplicationId = $TestApplicationId
+    testApplicationOid = "$TestApplicationOid"
+}
+
+if ($TenantId) {
+    $templateParameters.Add('tenantId', $TenantId)
+}
+if ($TestApplicationSecret) {
+    $templateParameters.Add('testApplicationSecret', $TestApplicationSecret)
+}
+if ($AdditionalParameters) {
+    $templateParameters += $AdditionalParameters
+}
+
+# Try to detect the shell based on the parent process name (e.g. launch via shebang).
+$shell, $shellExportFormat = if (($parentProcessName = (Get-Process -Id $PID).Parent.ProcessName) -and $parentProcessName -eq 'cmd') {
+    'cmd', 'set {0}={1}'
+} elseif (@('bash', 'csh', 'tcsh', 'zsh') -contains $parentProcessName) {
+    'shell', 'export {0}={1}'
+} else {
+    'PowerShell', '$env:{0} = ''{1}'''
+}
+
+foreach ($templateFile in $templateFiles) {
+    # Deployment fails if we pass in more parameters than are defined.
+    Write-Verbose "Removing unnecessary parameters from template '$templateFile'"
+    $templateJson = Get-Content -LiteralPath $templateFile | ConvertFrom-Json
+    $templateParameterNames = $templateJson.parameters.PSObject.Properties.Name
+
+    $templateFileParameters = $templateParameters.Clone()
+    foreach ($key in $templateParameters.Keys) {
+        if ($templateParameterNames -notcontains $key) {
+            Write-Verbose "Removing unnecessary parameter '$key'"
+            $templateFileParameters.Remove($key)
+        }
+    }
+
+    $preDeploymentScript = $templateFile | Split-Path | Join-Path -ChildPath 'test-resources-pre-deploy.ps1'

```

Comment:
I think the name is fine. Did you consider dropping `-deploy` suffix? `test-resources-pre/post`

Reply:
That was actually my first names. :) I was afraid they were 1) non-descript, and 2) hard to expand on should we want to add similar stuff later. But that's probably very unlikely.